src/decompose.rs

```diff
@@ -69,63 +74,72 @@ impl<I: Iterator<Item=char>> Iterator for Decompositions<I> {
     fn next(&mut self) -> Option<char> {
         use self::DecompositionType::*;
 
-        match self.buffer.first() {
-            Some(&(c, 0)) => {
-                self.sorted = false;
-                self.buffer.remove(0);
-                return Some(c);
-            }
-            Some(&(c, _)) if self.sorted => {
-                self.buffer.remove(0);
+        match self.buffer.pop_front() {
+            Some((c, _)) => {
                 return Some(c);
             }
-            _ => self.sorted = false
-        }
+            _ => {}
+        };
 
-        if !self.sorted {
-            for ch in self.iter.by_ref() {
-                let buffer = &mut self.buffer;
-                let sorted = &mut self.sorted;
-                {
-                    let callback = |d| {
-                        let class =
-                            super::char::canonical_combining_class(d);
-                        if class == 0 && !*sorted {
-                            canonical_sort(buffer);
-                            *sorted = true;
-                        }
-                        buffer.push((d, class));
-                    };
-                    match self.kind {
-                        Canonical => {
-                            super::char::decompose_canonical(ch, callback)
+        let mut non_starter_count = 0;
+        let stream_safe = self.stream_safe;
+
+        while self.iter.peek().is_some() {
+            let ch = self.iter.next().unwrap();
+            let ch_next = self.iter.peek();
+
+            let buffer = &mut self.buffer;
+            {
+                let callback = |d| {
+                    let class = super::char::canonical_combining_class(d);
+
+                    buffer.push_back((d, class));
+
+                    if stream_safe {
+                        if class == 0 {
+                            non_starter_count = 0;
+                        } else {
+                            non_starter_count += 1;
                         }
-                        Compatible => {
-                            super::char::decompose_compatible(ch, callback)
+
+                        if non_starter_count == 30 {
+                            match ch_next {
+                                Some(c) => {
+                                    if super::char::canonical_combining_class(*c) != 0 {
```

Comment:
does the algorithm here fundamentally need peekable?  can we "charge" this operation to the 31st character (i.e. `ch_next`) instead?

```
if class == 0 {
    buffer.push_back((d, class));
    non_starter_count = 0;
else if non_starter_count < 31 {
    buffer.push_back((d, class));
    non_starter_count += 1;
} else {
    canonical_sort(buffer);
    buffer.push_back((COMBINING_GRAPHEME_JOINER, 0));
    buffer.push_back((d, class));
    non_starter_count = 0;
}
```

Reply:
It doesn't fundamentally, no. Once you've added the `COMBINING_GRAPHEME_JOINER` it's safe to sort and emit the codepoints that are in the buffer before the CGJ, but not the one that comes after it in this case.

Not using Peekable basically means you're putting the peeked character into `self.buffer` instead, but you get to save 4 bytes.

Reply:
(Although I realize that the current implementation here doesn't actually take advantage of this, which is the whole point of the extension. Fixing...)